### PR TITLE
Fix calling ObReferenceObjectByHandle at DISPATCH

### PR DESCRIPTION
**Pull request by saxena-anurag**
### Issue
`ObReferenceObjectByHandle` can only be called at PASSVIE, but it was being called at DISPATCH. The map-in-map code is calling this API with spin lock held. This was not caught earlier due to missing map-in-map test cases in API test.

### Fix
This PR changes code to not hold the lock when calling `ObReferenceObjectByHandle`

## Testing

Added new test cases in api_test to cover this scenario.

## Documentation

NA

### CODE REVIEW

File: tests/api_test/api_test.cpp [hunk 1]
```diff
@@ -314,3 +314,83 @@ TEST_CASE("divide_by_zero_jit", "[divide_by_zero]") { divide_by_zero_test_km(EBP
...
 #endif
+
+void
+_test_nested_maps(bpf_map_type type)
+{
+    // Create 2 inner maps.
+    fd_t inner1 = bpf_create_map(BPF_MAP_TYPE_ARRAY, sizeof(uint32_t), sizeof(uint32_t), 1, 0);
+    REQUIRE(inner1 > 0);
+
+    // Create outer map.
```
Comment (shankarseal) [inline tests/api_test/api_test.cpp:325]: consider moving these 2 lines above line 321.

File: libs/execution_context/ebpf_maps.c [hunk 2]
```diff
@@ -496,20 +496,12 @@ _check_value_type(_In_ const ebpf_core_map_t* outer_map, _In_ const ebpf_core_ob
...
-// Validate that a value handle is appropriate for this map,
-// and if so, return a pointer to the object with that handle.
+// Validate that a value object is appropriate for this map.
 static ebpf_result_t
-_get_map_value_object(
-    _In_ const ebpf_core_map_t* map,
-    ebpf_handle_t value_handle,
-    ebpf_object_type_t value_type,
-    _Outptr_ ebpf_core_object_t** value_object_result)
+_validate_map_value_object(
```
Comment (shankarseal) [inline libs/execution_context/ebpf_maps.c:501]: if this function requires the object lock to be held, please add required SAL annotations.

File: tests/api_test/api_test.cpp (hunk 1 above)
Comment (dthaler) [reply tests/api_test/api_test.cpp:325]: You can't since inner1 is used in line 326.

File: tests/api_test/api_test.cpp [hunk 3]
```diff
@@ -314,3 +314,83 @@ TEST_CASE("divide_by_zero_jit", "[divide_by_zero]") { divide_by_zero_test_km(EBP
...
+    struct bpf_object* object = nullptr;
+    fd_t program_fd;
+
+    result =
+        _program_load_helper("tail_call_multiple.o", &EBPF_PROGRAM_TYPE_XDP, EBPF_EXECUTION_ANY, &object, &program_fd);
+    REQUIRE(result == EBPF_SUCCESS);
+
+    REQUIRE(program_fd > 0);
+
+    // Setup tail calls.
```
Comment (dthaler) [inline tests/api_test/api_test.cpp:370]:
```suggestion
    // Set up tail calls.
```

**Review by dthaler — APPROVED** [review]

File: libs/execution_context/ebpf_maps.c (hunk 2 above)
Comment (saxena-anurag) [reply libs/execution_context/ebpf_maps.c]: Added.

File: libs/execution_context/ebpf_maps.c [hunk 4]
```diff
@@ -496,20 +496,12 @@ _check_value_type(_In_ const ebpf_core_map_t* outer_map, _In_ const ebpf_core_ob
...
-// and if so, return a pointer to the object with that handle.
-static ebpf_result_t
-_get_map_value_object(
-    _In_ const ebpf_core_map_t* map,
-    ebpf_handle_t value_handle,
-    ebpf_object_type_t value_type,
-    _Outptr_ ebpf_core_object_t** value_object_result)
+// Validate that a value object is appropriate for this map.
+static _Requires_lock_held_(object_map->lock) ebpf_result_t _validate_map_value_object(
+    _In_ const ebpf_core_object_map_t* object_map, ebpf_object_type_t value_type, _In_ ebpf_core_object_t* value_object)
```
Comment (shankarseal) [inline libs/execution_context/ebpf_maps.c:501]: `value_object` should be `const` as well.

File: libs/execution_context/ebpf_maps.c [hunk 5]
```diff
@@ -1071,15 +1074,14 @@ _update_hash_map_entry_with_handle(
...
         // The hash table is already full.
         result = EBPF_OUT_OF_SPACE;
         goto Done;
     }
 
-    result = _get_map_value_object(map, value_handle, value_type, &value_object);
+    result = _validate_map_value_object(object_map, value_type, value_object);
     if (result != EBPF_SUCCESS) {
         goto Done;
     }
```
Comment (shankarseal) [inline libs/execution_context/ebpf_maps.c:1087]: there's a call to `ebpf_object_release_reference` within lock. Is that OK? In `_update_array_map_entry_with_handle` you changed to call it outside lock.

File: tests/api_test/api_test.cpp (hunk 1 above)
Comment (shankarseal) [reply tests/api_test/api_test.cpp:325]: Sorry, I missed that. However, couldn't you call the `create_map_in_map` API with fd to the template map that should be defined in the ELF file. Do we have a test for that?

**Review by shankarseal — DISMISSED** [review]

File: libs/execution_context/ebpf_maps.c (hunk 5 above)
Comment (saxena-anurag) [reply libs/execution_context/ebpf_maps.c:1087]: It should not make any difference. But still reverted to the original behavior (releasing reference under lock).

File: libs/execution_context/ebpf_maps.c (hunk 4 above)
Comment (saxena-anurag) [reply libs/execution_context/ebpf_maps.c]: Fixed. Also removed const from "object_map" as it is being modified in this function.

**Review by dthaler — APPROVED** [review]

**Review by shankarseal — APPROVED** [review]

**Pull request merged**